Refactor eda and baseline model

marielledado opened the pull request:
Feedback welcome! I've never actually done this before so I'd love to know how this can be improved.

I realized that I didn't actually create a script for the baseline model per se...just a bunch of helper functions for generic data prep and modeling. So if you have better ideas, please let me know!

Here are the most important files:

/emissionscheck_alb/
- `data_prep.py` - all helper functions for data cleaning and data prep (e.g., encoding)
- `baseline.py` - function for running logistic regression and SVC

exploration/01_eda_minimal_model/
- `week3_eda.ipynb` - refactored EDA notebook from week 3 which uses functions from `data_prep.py`
- `week3_baseline_model.ipynb` - refactored baseline model notebook from week 3 which uses functions from `data_prep.py` and `baseline.py`

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+        DataFrame of rows with desired values `'P'` or `'F'` in target variable 
+    '''
+    # Mask to select only rows with desired target values
+    overall_result_mask = (df['OVERALL_RESULT'] == 'P') | (df['OVERALL_RESULT'] == 'F')
```
this can be shortened to:
```suggestion
    df = df[df.OVERALL_RESULT.isin(['P', 'F'])]

```

marielledado replied:
Done

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+    # Mask to select only rows with desired target values
+    overall_result_mask = (df['OVERALL_RESULT'] == 'P') | (df['OVERALL_RESULT'] == 'F')
+    
+    return df[overall_result_mask]
```
```suggestion
    return df

```

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+    # add +1 to df['CAR_AGE']
+    df['CAR_AGE'] += 1
+    
+    return df
```
git is complaining, apparently it want a newline at the end of a script
```suggestion
    return df
    
```

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+    Adding +1 to car age to avoid negative values (many cars are released a year prior to year in car model)
+    
+    '''
+    # convert df['TEST_EDATE'] to datetime
```
remove comment
```suggestion
```

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+def keep_pass_fail(df):
```
once these functions are part of a larger class, we might want to start their names with an underscore, to signal that they are helper functions and not meant to be used outside of the class

marielledado replied:
Sounds good to me. I've never actually defined classes in Python so I'd love to learn how this is done.

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+    # convert df['TEST_EDATE'] to datetime
+    df['CAR_AGE'] = pd.to_datetime(df['TEST_EDATE']).dt.year - df['MODEL_YEAR']
+    
+    # add +1 to df['CAR_AGE']
```
you already say why you are doing this in the docstring, remove comment
```suggestion
```

marielledado replied:
I removd the docstring note and kept the comment

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -0,0 +1,46 @@
+    
+    Parameters
+    ----------
+    df : pandas.DataFrame
```
this is very wordy. input with type and output with type should be enough

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -15,36 +15,59 @@ def keep_pass_fail(df):
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
```
are you importing a dataframe or a series? docstring says df, but code suggests series

marielledado replied:
The function accepts a DataFrame, makes a copy of it, and then takes the relevant columns.

izzbizz replied:
hmm, am I looking at an old version again? in here, the function takes a series, modifies it and returns it. the comment says otherwise
![image](https://user-images.githubusercontent.com/34165505/84576666-d5275800-adb6-11ea-9134-b220266b766f.png)

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -15,36 +15,59 @@ def keep_pass_fail(df):
+    pandas.DataFrame
+        DataFrame with encoded `F` and `P` target variable 
+    '''
+    # make copy of df
```
remove all comments
```suggestion
```

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -15,36 +15,59 @@ def keep_pass_fail(df):
+    '''
+    # make copy of df
+
+    # Mask to select only rows with desired target values
```
remove
```suggestion
```

izzbizz commented on emissionscheck_alb/data.py:
```diff
@@ -1,3 +1,6 @@
+import pandas as pd
+
+
```
one newline is enough
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+
+    if model == "logreg":
+
+        from sklearn.linear_model import LogisticRegression
```
import all your packages at the top of your script
```suggestion
```

marielledado replied:
Done

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+import pandas as pd
+import numpy as np
+
```
do all your imports here, it makes the code more readable (and once the script gets longer, you have a far better overview over what you're importing)
```suggestion
    from sklearn.pipeline import make_pipeline
    from sklearn.preprocessing import MinMaxScaler
    from sklearn.metrics import roc_auc_score, plot_confusion_matrix, classification_report
    from sklearn.model_selection import cross_val_score
    from sklearn.linear_model import LogisticRegression
    from sklearn.svm import SVC
```

marielledado replied:
Done!

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    Run logistic regression or SVC  
+    
+    """
+    from sklearn.pipeline import make_pipeline
```
see above
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    
+    """
+    from sklearn.pipeline import make_pipeline
+    from sklearn.preprocessing import MinMaxScaler
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    """
+    from sklearn.pipeline import make_pipeline
+    from sklearn.preprocessing import MinMaxScaler
+    from sklearn.metrics import roc_auc_score, plot_confusion_matrix, classification_report
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    from sklearn.pipeline import make_pipeline
+    from sklearn.preprocessing import MinMaxScaler
+    from sklearn.metrics import roc_auc_score, plot_confusion_matrix, classification_report
+    from sklearn.model_selection import cross_val_score
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    elif model == "svc":
+
+
+        from sklearn.svm import SVC
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    print(f'Cross-validation scores: {cv_score}')
+    print(f'Mean cross-validation score: {mean_cv_score}')
+
+
```
only one newline
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+        from sklearn.svm import SVC
+        svc = SVC(class_weight=class_weight)
+
+        clf_pipeline = make_pipeline(MinMaxScaler(), svc)
```
i think it's really cool how you are modularizing the use of the models. i was thinking we could have a class, and when we initialize it, we pass the name of the model and variables we want to use.

marielledado replied:
I'm also open to that! I don't exactly know how that works so if you have time/are interested feel free to refactor this script to do that. Or maybe implement a script like this for future models.

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+
+    elif model == "svc":
+
+
```
only one newline
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+        clf_pipeline = make_pipeline(MinMaxScaler(), svc)
+
+
+    # fit model
```
only comment on things that are not obvious
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    # fit model
+    fitted_model = clf_pipeline.fit(features_train, target_train)
+
+    # predict labels
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    # predict labels
+    target_pred = clf_pipeline.predict(features_test)
+
+    # predict predicted probabilities
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    # predict predicted probabilities
+    target_pred_proba = clf_pipeline.predict_proba(features_test)
+
+    # classification report
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    target_pred_proba = clf_pipeline.predict_proba(features_test)
+
+    # classification report
+    print(classification_report(target_test, target_pred))
```
if you print this, maybe print a description too, like you are doing with the AUC score

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    # classification report
+    print(classification_report(target_test, target_pred))
+
+    # confusion matrix
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    # confusion matrix
+    plot_confusion_matrix(clf_pipeline, features_test, target_test)
+
+    # calculate ROC
```
```suggestion
```

izzbizz commented on emissionscheck_alb/baseline_model.py:
```diff
@@ -0,0 +1,59 @@
+    roc_auc = roc_auc_score(target_test, target_pred_proba[:, 1], max_fpr=max_fpr)
+    print(f'ROC AUC Score is {roc_auc}')
+    
+    # cross val score
```
```suggestion
```

izzbizz commented on emissionscheck_alb/data_prep.py:
```diff
@@ -0,0 +1,161 @@
+    """
+    from sklearn.model_selection import train_test_split
+
+    features_train, features_test, target_train, target_test = train_test_split(
```
is there a  reason why you don't use X and y? since they are so widely used, i find them easier to read (and shorter!) than features and targets

marielledado replied:
Personally I find `features` and `target` more descriptive and less confusing.

izzbizz commented on emissionscheck_alb/data_prep.py:
```diff
@@ -0,0 +1,161 @@
+
+    features_encoded = pd.concat([df_working_copy[cont_features], features_dummy], axis=1)
+
+    from sklearn.model_selection import train_test_split
```
import at the top

izzbizz commented on emissionscheck_alb/data_prep.py:
```diff
@@ -0,0 +1,161 @@
+
+    return features_train, features_test, target_train, target_test
+
+def encode_split_data(cat_features, cont_features, data, target_series, train_size, random_state=None):
```
why are you doing both the encoding and the splitting in this function? you already have a splitting function above , so you could just make this one do the encoding and then pass the output to the splitting, otherwise, cool function :)

marielledado replied:
I think it's because I wanted to avoid having call another splitting function after encoding categorical features.

izzbizz commented on emissionscheck_alb/data_prep.py:
```diff
@@ -0,0 +1,161 @@
+    df_working_copy = data.copy()
+    
+    # convert categorical columns into `category` dtype
+    for column_name in df_working_copy[cat_features].columns:
```
```suggestion
    for column_name in cat_features:

```

izzbizz commented on emissionscheck_alb/data_prep.py:
```diff
@@ -0,0 +1,161 @@
+    cat_features : list of str
+        List of column names in `data` containing categorical features
+    
+    cont_features : list of str
```
you don't need to pass a list of continuous features, since what you are concatenating later is just the complement of your categorical columns. so you can do something like: 
`cont_features = [col for col in data.columns if not col in cat_features]`

marielledado replied:
Good point! Will implement this. This assumes though that you first created a subset of your original `DataFrame` with only the columns you're interested in.

marielledado commented:
Thanks again @izzbizz for your valuable comments! I implemented many of your suggestions, especially the ones that shorten/simplify the code. Here's a quick overview of the changes:

- Imports at the very beginning of the script
- Fewer comments, removed the ones that are fairly obvious
- Shortened some of the docstrings. I still prefer this format even if it's a bit longer than just description + data type. Besides, it follows the same conventions you see in docstrings of many Python packages. 

If you have time in the coming week(s), I would love to see how you would initialize a class and pass the models/variables.
